Fix: Theme search on mobile

enejb opened the pull request:
#### Changes proposed in this Pull Request

* Do not attempt to scroll to the top on mobile only. 
* Do not use page() call since it also moves the page to the top when there is no search

#### Testing instructions
Go to http://calypso.localhost:3000/themes/example.com

* Use the search bar. On mobile and desktop) Notice that it works as expected. 
* Scrolls to the top on search typing. The search bar doesn't move around as you type. 

Fixes https://github.com/Automattic/wp-calypso/issues/58149


Before:

https://user-images.githubusercontent.com/115071/143321584-d2b31dee-78ac-4a9e-af5e-74a287ffa0e7.mp4

https://user-images.githubusercontent.com/115071/143321593-61bd16eb-592e-4f76-8d96-6ff5c54ebb4e.mp4


After:

https://user-images.githubusercontent.com/115071/143321574-f187c24d-1491-4dc3-b421-750e945e88d5.mp4


https://user-images.githubusercontent.com/115071/143321599-7a53090e-add3-4b7c-b1cc-906e55e023b5.mp4

enejb committed:
Fix: Theme search on mobile

enejb committed:
Move the search bar to the top.

enejb committed:
Make sure that there is always enough height to be position with the correct height

tyxla commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -159,7 +159,10 @@ class ThemeShowcase extends Component {
 
 	scrollToSearchInput = () => {
 		if ( ! this.props.loggedOutComponent && this.scrollRef && this.scrollRef.current ) {
+			const yOffset = -55;
```
That'll most certainly need a comment so we can know where it came from.

enejb replied:
Comment added.

tyxla commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -376,7 +379,7 @@ class ThemeShowcase extends Component {
 					/>
 				) }
-				<div className="themes__content">
+				<div className="themes__content" ref={ this.scrollRef }>
```
So weird that this ref wasn't here before. What were we scrolling before?

enejb replied:
I think the scrollRef was missed during a code refactor.

tyxla reviewed (commented):
This is a bit better, but I'm still seeing a scroll if I initiate a theme search on mobile while scrolled further down. 

Should we perhaps disable the `scrollToSearchInput` for mobile devices completely? What's its purpose in the first place?

enejb committed:
Add a comment

enejb commented:
> Should we perhaps disable the scrollToSearchInput for mobile devices completely? What's its purpose in the first place?

I tried that approach as well first but one issue that I was running into still is the view being scrolled as you typed. This happens because on each character the URL changes and so the user gets taken to the top of the page because of it. See https://github.com/Automattic/wp-calypso/pull/58318/files#diff-5e6c175ac60d9ed48935454e0bcf1861c5d5893c33a08edb946162d2e0dea550L179


I think the scrollToSearchInput makes sure that the input is always visible. This works fine on the desktop but it doesn't work well on mobile because the search box can be hidden by the keyboard and things move around quite a lot as well since users usually type way slower on mobile devices. ( at least I do )

enejb commented:
Thanks for taking a look!

When you have a moment can you take another look?

I think this PR is definitely an improvement to the mobile side let me know if there is anything that you want to change :)

tyxla commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -159,7 +159,10 @@ class ThemeShowcase extends Component {
 
 	scrollToSearchInput = () => {
 		if ( ! this.props.loggedOutComponent && this.scrollRef && this.scrollRef.current ) {
+			const yOffset = -55; // A number that takes the Adminbar as well as the screen options into account on mobile view ports.
```
Ideally, we could dynamically take the height of these elements and calculate it before passing it here. That'll ensure this continues to work even if we do some style modifications in the future.

enejb replied:
Great point

tyxla commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -159,7 +159,10 @@ class ThemeShowcase extends Component {
+			const yOffset = -55; // A number that takes the Adminbar as well as the screen options into account on mobile view ports.
 			this.scrollRef.current.scrollIntoView();
+			const y = this.scrollRef.current.getBoundingClientRect().top + window.pageYOffset + yOffset;
+			window.scrollTo( { top: y } );
```
Could we do this only for mobile? It feels odd that we're scrolling for desktop too, when there's plenty of space for everything.

tyxla reviewed (commented):
I think this is ready to go, but just wanted to ask you a couple of questions before we 🚢  it.

enejb committed:
Calcualte the height offset dynamically + only scrollTo on mobile

600 is picked because that is when the height of the Theme Info box changes a lot.

enejb commented:
Thanks for the suggestion @tyxla 
This should be working more as expected on desktop and mobile now. 

Let me know if you have any more questions :)

tyxla commented on client/components/screen-options-tab/index.js:
```diff
@@ -92,7 +92,12 @@ const ScreenOptionsTab = ( { wpAdminPath } ) => {
-		<div className="screen-options-tab" ref={ ref } data-testid="screen-options-tab">
+		<div
+			className="screen-options-tab"
+			id="screen-options-tab-id"
```
Theoretically, we can have multiple instances of this component in the same render. That would cause the same ID to be repeated multiple times, which isn't ideal. 

Can we use the classname instead?

enejb replied:
You are right!

tyxla commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -159,7 +159,20 @@ class ThemeShowcase extends Component {
+			}
+			const headerHeight = document.getElementById( 'header' )?.getBoundingClientRect().height;
+			const screenOptionTab = document
+				.getElementById( 'screen-options-tab-id' )
```
We could select the first element by classname with `document.querySelector()`. Or we could use `document.getElementsByClassName()` too. Up to you.

enejb replied:
:+1:

tyxla commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -159,7 +159,20 @@ class ThemeShowcase extends Component {
+			if ( window.innerWidth > 600 ) {
+				return;
+			}
+			const headerHeight = document.getElementById( 'header' )?.getBoundingClientRect().height;
```
Same for the `id="header"` here - could we select by classname and use the first element with `.masterbar` instead?

enejb replied:
Good idea!

tyxla commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -159,7 +159,20 @@ class ThemeShowcase extends Component {
+				.getElementById( 'screen-options-tab-id' )
+				?.getBoundingClientRect().height;
+
+			const yOffset = -( headerHeight + screenOptionTab ); // A number that takes the Adminbar as well as the screen options into account on mobile view ports.
```
Thanks, that makes a lot sense now 👍 

I'd suggest shortening the comment a bit if possible to keep the line shorter, maybe something like:

```suggestion
			const yOffset = -( headerHeight + screenOptionTab ); // Total height of admin bar and screen options on mobile.
```

tyxla commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -159,7 +159,20 @@ class ThemeShowcase extends Component {
 		if ( ! this.props.loggedOutComponent && this.scrollRef && this.scrollRef.current ) {
-			this.scrollRef.current.scrollIntoView();
+			// If you are a larger screen where the theme info is displayed horizontally.
+			if ( window.innerWidth > 600 ) {
```
This seems a bit raw; does it make sense to use one of our viewport packages to achieve the same?

enejb replied:
I looked for a 600 px viewport but couldn't find it in the calypso library but wasn't able to find it. 
I wanted to target 600 px specifically since that is the viewport when the theme info box starts to be more horizontal. 
(Image next to the text) so the issue of the keyboard hiding the search box is not so prevalent.

tyxla replied:
Gutenberg's viewport package has a 600px viewport FWIW: https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/viewport/src/index.js#L22

I don't think we're using it in Calypso yet, but it might be a good opportunity.

enejb committed:
Update the comment.

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

enejb committed:
Use classname select query to find the elements

this is done since more then a few different of the same component could be present

tyxla reviewed (approved):
This LGTM and tests well! Nice improvement for mobile 👍  ✨ 

Just left one optional suggestion regarding the hardcoded breakpoint size.

enejb commented:
Thanks for the review @tyxla 

I merged this PR and was planning on making viewport changes in a different PR. 

But in order to do that will I first need to make a Gutenberg pr to export the variables? Do you think it is worth doing this?

tyxla commented:
> But in order to do that will I first need to make a Gutenberg pr to export the variables? Do you think it is worth doing this?

It doesn't seem to be worth it. I was just wondering why we need to export them in the first place? Can't we just do something like:

```
if( isViewportMatch( '< small' ) ) {
	return;
}
```
